<center><p float="center">
  <img src="https://upload.wikimedia.org/wikipedia/commons/e/e9/4_RGB_McCombs_School_Brand_Branded.png" width="600" height="200"/>
</p></center>

<h1><center><font size=8>Bancassurance</center></font></h1>
<h1><center>Analitis, Iraklis - Project 3</center></h1>

# **Introduction**

## **Description**
### *Background and Context*

Best insurance company and My Bank have set up a Bancassurance(Bancassurance is a relationship between a bank and an insurance company), now using the data of liability customers of My Bank, The Best insurance company wants to convert customers with both a life insurance policy and an account in My bank to loan customers(taking a loan against a life insurance policy)
A campaign that the company ran last year for liability customers showed a healthy conversion rate of over 12.56% success. You are provided with data of customers who have an account in My bank and life insurance policy in the Best insurance company 
You as a data scientist at the Best insurance company have to build a model to identify the positively responding customers who have a higher probability of purchasing the insurance. This will increase the success ratio and reduce the cost of the campaign.


### *Objective*

* To predict whether a liability customer will buy a loan or not.
* Which variables are most significant for making predictions.
* Which segment of customers should be targeted more.


## **Data Dictionary**

* CUST_ID: Unique Customer ID
* Target: Field - 1: Responder, 0: Non-Responder
* Age: Customer Age in years
* Gender: Male / Female / Other
* Balance: Monthly Average Balance
* Occupation: Professional / Salaried / Self Employed / SelfEmployed Non Professional.
* Age_BKT: Age Bucket
* SCR: Marketing Score
* HOLDING_PERIOD: Duration in days to hold the money
* ACC_TYPE: Account Type: Current Account / Saving Account
* ACC_OP_DATE: Account Open Date
* LEN_OF_RLTN_IN_MNTH: Length of Relationship in Months
* NO_OF_L_CR_TXNS: Number of Credit Transactions
* NO_OF_L_DR_TXNS: Number of Debit Transactions
* TOT_NO_OF_L_TXNS: Total Number of Transactions
* NO_OF_BR_CSH_WDL_DR_TXNS: Branch Cash Withdrawal Debit Transactions
* NO_OF_ATM_DR_TXNS: Number of ATM Debit Transactions
* NO_OF_NET_DR_TXNS: Number of Net Banking Debit Transactions
* NO_OF_MOB_DR_TXNS: Number of Mobile Banking Debit Transactions
* NO_OF_CHQ_DR_TXNS: Number of Cheque Debit Transactions
* FLG_HAS_CC: Has Credit Card - 1: Yes, 0: No
* AMT_ATM_DR: Amount Withdrawn from ATM
* AMT_BR_CSH_WDL_DR: Amount cash withdrawn from Branch
* AMT_CHQ_DR: Amount debited by Cheque Transactions
* AMT_NET_DR: Amount debited by Net Transactions
* AMT_MOB_DR: Amount debited by Mobile Transactions
* AMT_L_DR: Total Amount Debited
* FLG_HAS_ANY_CHGS: Flag: Has any banking charges
* AMT_OTH_BK_ATM_USG_CHGS: Charges for using Other Bank ATM
* AMT_MIN_BAL_NMC_CHGS: Charges for not maintaining Min Balance
* NO_OF_IW_CHQ_BNC_TXNS: Charges for Inward Cheque Bounce
* NO_OF_OW_CHQ_BNC_TXNS: Charges for Outward Cheque Bounce
* AVG_AMT_PER_ATM_TXN: Average Amount per ATM Transaction
* AVG_AMT_PER_CSH_WDL_TXN: Average Amount Per Cash Withdrawal
* AVG_AMT_PER_CHQ_TXN: Average Amount Per Cheque Transaction
* AVG_AMT_PER_NET_TXN: Average Amount per Net banking Transaction
* AVG_AMT_PER_MOB_TXN: Average Amount per Mobile Transaction
* FLG_HAS_NOMINEE: Flag: Has Nominee - 1: Yes, 0: No
* FLG_HAS_OLD_LOAN: Flag: Has any earlier loan - 1: Yes, 0: No
* random: Random Number

#**Initial Assessment of Data**

## Import necessary libraries

In [1]:
import warnings

warnings.filterwarnings("ignore")
from statsmodels.tools.sm_exceptions import ConvergenceWarning

warnings.simplefilter("ignore", ConvergenceWarning)

# Libraries to help with reading and manipulating data

import pandas as pd
import numpy as np

# Library to split data
from sklearn.model_selection import train_test_split

# libaries to help with data visualization
import matplotlib.pyplot as plt
import seaborn as sns

# Removes the limit for the number of displayed columns
pd.set_option("display.max_columns", None)
# Sets the limit for the number of displayed rows
pd.set_option("display.max_rows", 200)


# To build model for prediction
from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier
from sklearn import tree

# To tune different models
from sklearn.model_selection import GridSearchCV


# To get diferent metric scores
from sklearn.metrics import (
    f1_score,
    accuracy_score,
    recall_score,
    precision_score,
    confusion_matrix,
    roc_auc_score,
    plot_confusion_matrix,
    precision_recall_curve,
    roc_curve,
    make_scorer,
)

# For pandas profiling
from pandas_profiling import ProfileReport

## Data Preprocessing

In [2]:
#Mount drive
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [3]:
#Read .csv file to construct DataFrame object
Bank = pd.read_csv('/content/drive/MyDrive/ColabNotebooks/UTAustinAIMLProjects/Project3/MyBank.csv')

In [4]:
# copying data to another varaible to avoid any changes to original data
data = Bank.copy()

### View the first and last 5 rows of the dataset.

In [5]:
data.head()

CUST_ID  TARGET  AGE GENDER     BALANCE OCCUPATION AGE_BKT  SCR  \
0   C7927       0   27      M     3383.75   SELF-EMP   26-30  776   
1   C6877       0   47      M   287489.04        SAL   46-50  324   
2  C19922       0   40      M    18216.88   SELF-EMP   36-40  603   
3   C8183       0   53      M    71720.48        SAL     >50  196   
4  C12123       0   36      M  1671622.89       PROF   36-40  167   

   HOLDING_PERIOD ACC_TYPE ACC_OP_DATE  LEN_OF_RLTN_IN_MNTH  NO_OF_L_CR_TXNS  \
0              30       SA   3/23/2005                  146                7   
1              28       SA    10-11-08                  104                8   
2               2       SA   4/26/2012                   61               10   
3              13       CA    07-04-08                  107               36   
4              24       SA  12/29/2001                  185               20   

   NO_OF_L_DR_TXNS  TOT_NO_OF_L_TXNS  NO_OF_BR_CSH_WDL_DR_TXNS  \
0                3                10                         0   
1                2                10                         0   
2                5                15                         1   
3               14                50                         4   
4                1                21                         1   

   NO_OF_ATM_DR_TXNS  NO_OF_NET_DR_TXNS  NO_OF_MOB_DR_TXNS  NO_OF_CHQ_DR_TXNS  \
0                  1                  2                  0                  0   
1                  1                  1                  0                  0   
2                  1                  1                  0                  2   
3                  2                  3                  1                  4   
4                  0                  0                  0                  0   

   FLG_HAS_CC  AMT_ATM_DR  AMT_BR_CSH_WDL_DR  AMT_CHQ_DR  AMT_NET_DR  \
0           0       13100                  0           0    973557.0   
1           0        6600                  0           0    799813.0   
2           0       11200             561120       49320    997570.0   
3           0       26100             673590       60780    741506.0   
4           0           0             808480           0         0.0   

   AMT_MOB_DR   AMT_L_DR  FLG_HAS_ANY_CHGS  AMT_OTH_BK_ATM_USG_CHGS  \
0           0   986657.0                 0                        0   
1           0   806413.0                 1                        0   
2           0  1619210.0                 1                        0   
3       71388  1573364.0                 0                        0   
4           0   808480.0                 0                        0   

   AMT_MIN_BAL_NMC_CHGS  NO_OF_IW_CHQ_BNC_TXNS  NO_OF_OW_CHQ_BNC_TXNS  \
0                     0                      0                      0   
1                     0                      0                      0   
2                     0                      0                      1   
3                     0                      0                      0   
4                     0                      0                      0   

   AVG_AMT_PER_ATM_TXN  AVG_AMT_PER_CSH_WDL_TXN  AVG_AMT_PER_CHQ_TXN  \
0              13100.0                      0.0                  0.0   
1               6600.0                      0.0                  0.0   
2              11200.0                 561120.0              24660.0   
3              13050.0                 168397.5              15195.0   
4                  0.0                 808480.0                  0.0   

   AVG_AMT_PER_NET_TXN  AVG_AMT_PER_MOB_TXN  FLG_HAS_NOMINEE  \
0            486778.50                  0.0                1   
1            799813.00                  0.0                1   
2            997570.00                  0.0                1   
3            247168.67              71388.0                1   
4                 0.00                  0.0                1   

   FLG_HAS_OLD_LOAN    random  
0                 1  0.000011  
1                 0  0.000111  
2        

In [6]:
data.tail()

CUST_ID  TARGET  AGE GENDER    BALANCE OCCUPATION AGE_BKT  SCR  \
19995   C4520       0   41      F  181029.01       SENP   41-45  598   
19996  C10147       0   37      F   46624.98   SELF-EMP   36-40  299   
19997   C2392       0   30      M  142433.53       PROF   26-30  722   
19998  C14659       0   46      M  394654.75        SAL   46-50  234   
19999  C12911       1   41      M    7152.60   SELF-EMP   41-45  561   

       HOLDING_PERIOD ACC_TYPE ACC_OP_DATE  LEN_OF_RLTN_IN_MNTH  \
19995              15       SA    09-04-03                  165   
19996              11       SA    05-03-13                   49   
19997              11       SA    11-01-07                  115   
19998               4       SA   8/22/2006                  129   
19999              12       SA   6/21/2014                   35   

       NO_OF_L_CR_TXNS  NO_OF_L_DR_TXNS  TOT_NO_OF_L_TXNS  \
19995                4                3                 7   
19996               17               14                31   
19997               20               16                36   
19998               11                6                17   
19999               10                6                16   

       NO_OF_BR_CSH_WDL_DR_TXNS  NO_OF_ATM_DR_TXNS  NO_OF_NET_DR_TXNS  \
19995                         1                  0                  0   
19996                         4                  2                  3   
19997                         8                  2                  0   
19998                         1                  1                  0   
19999                         1                  1                  0   

       NO_OF_MOB_DR_TXNS  NO_OF_CHQ_DR_TXNS  FLG_HAS_CC  AMT_ATM_DR  \
19995                  0                  2           0           0   
19996                  1                  4           0       17200   
19997                  1                  5           0       33400   
19998                  0                  4           1        7700   
19999                  0                  4           1       15100   

       AMT_BR_CSH_WDL_DR  AMT_CHQ_DR  AMT_NET_DR  AMT_MOB_DR   AMT_L_DR  \
19995             277050       59460         0.0           0   336510.0   
19996             705740       24300    902772.0       46829  1696841.0   
19997              83740       62210         0.0      132010   311360.0   
19998             195790       43550         0.0           0   247040.0   
19999             276520       61510         0.0           0   353130.0   

       FLG_HAS_ANY_CHGS  AMT_OTH_BK_ATM_USG_CHGS  AMT_MIN_BAL_NMC_CHGS  \
19995                 0                        0                     0   
19996                 0                        0                     0   
19997                 0                        0                     0   
19998                 0                        0                     0   
19999                 0                        0                     0   

       NO_OF_IW_CHQ_BNC_TXNS  NO_OF_OW_CHQ_BNC_TXNS  AVG_AMT_PER_ATM_TXN  \
19995                      0                      0                  0.0   
19996                      0                      0               8600.0   
19997                      0                      0              16700.0   
19998                      0                      0               7700.0   
19999                      0                      0              15100.0   

       AVG_AMT_PER_CSH_WDL_TXN  AVG_AMT_PER_CHQ_TXN  AVG_AMT_PER_NET_TXN  \
19995                 277050.0              29730.0                  0.0   
19996                 176435.0               6075.0             300924.0   
19997                  10467.5              12442.0                  0.0   
19998                 195790.0              10887.5                  0.0   
19999                 276520.0              15377.5                  0.0   

       AVG_AMT_PER_MOB_TXN  FLG_HAS_NOMINEE  FLG_HAS_OLD_LOAN    random  
19995                  0.0                1          

### Understand the shape of the dataset.

In [7]:
data.shape

(20000, 40)

* There are 20000 observations and 40 columns in the dataset

### Check the data types of the columns for the dataset.

In [8]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20000 entries, 0 to 19999
Data columns (total 40 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   CUST_ID                   20000 non-null  object 
 1   TARGET                    20000 non-null  int64  
 2   AGE                       20000 non-null  int64  
 3   GENDER                    20000 non-null  object 
 4   BALANCE                   20000 non-null  float64
 5   OCCUPATION                20000 non-null  object 
 6   AGE_BKT                   20000 non-null  object 
 7   SCR                       20000 non-null  int64  
 8   HOLDING_PERIOD            20000 non-null  int64  
 9   ACC_TYPE                  20000 non-null  object 
 10  ACC_OP_DATE               20000 non-null  object 
 11  LEN_OF_RLTN_IN_MNTH       20000 non-null  int64  
 12  NO_OF_L_CR_TXNS           20000 non-null  int64  
 13  NO_OF_L_DR_TXNS           20000 non-null  int64  
 14  TOT_NO

**Observations-**
* There are 6 columns with Dtype object, they may be categorical.
* There are no columns that appear to have missing values.

### Summary of the dataset.

In [9]:
data.describe().T

count           mean            std         min  \
TARGET                    20000.0       0.125600       0.331406    0.000000   
AGE                       20000.0      38.418150       9.577985   21.000000   
BALANCE                   20000.0  511362.192118  723592.961792    0.000000   
SCR                       20000.0     440.150300     249.887582  100.000000   
HOLDING_PERIOD            20000.0      14.955650       8.642634    1.000000   
LEN_OF_RLTN_IN_MNTH       20000.0     125.239300      54.953936   29.000000   
NO_OF_L_CR_TXNS           20000.0      12.348050      12.151921    0.000000   
NO_OF_L_DR_TXNS           20000.0       6.633700       7.675830    0.000000   
TOT_NO_OF_L_TXNS          20000.0      18.975400      17.759290    0.000000   
NO_OF_BR_CSH_WDL_DR_TXNS  20000.0       1.883000       2.263596    0.000000   
NO_OF_ATM_DR_TXNS         20000.0       1.028950       1.497239    0.000000   
NO_OF_NET_DR_TXNS         20000.0       1.172450       2.404246    0.000000   
NO_OF_MOB_DR_TXNS         20000.0       0.411750       2.009854    0.000000   
NO_OF_CHQ_DR_TXNS         20000.0       2.137550       2.384893    0.000000   
FLG_HAS_CC                20000.0       0.305400       0.460588    0.000000   
AMT_ATM_DR                20000.0   10990.015000   15304.891499    0.000000   
AMT_BR_CSH_WDL_DR         20000.0  378474.478000  331417.730129    0.000000   
AMT_CHQ_DR                20000.0  124520.005150  439379.541425    0.000000   
AMT_NET_DR                20000.0  237307.771646  319067.664858    0.000000   
AMT_MOB_DR                20000.0   22424.704400   49484.803631    0.000000   
AMT_L_DR                  20000.0  773716.974196  712659.173217    0.000000   
FLG_HAS_ANY_CHGS          20000.0       0.110600       0.313644    0.000000   
AMT_OTH_BK_ATM_USG_CHGS   20000.0       1.099500      14.219220    0.000000   
AMT_MIN_BAL_NMC_CHGS      20000.0       1.292000      14.764201    0.000000   
NO_OF_IW_CHQ_BNC_TXNS     20000.0       0.042750       0.203038    0.000000   
NO_OF_OW_CHQ_BNC_TXNS     20000.0       0.044400       0.206230    0.000000   
AVG_AMT_PER_ATM_TXN       20000.0    7408.839730    7210.168619    0.000000   
AVG_AMT_PER_CSH_WDL_TXN   20000.0  242236.478811  269837.890598    0.000000   
AVG_AMT_PER_CHQ_TXN       20000.0   25092.478384   49387.197536    0.000000   
AVG_AMT_PER_NET_TXN       20000.0  179059.029254  281706.855583    0.000000   
AVG_AMT_PER_MOB_TXN       20000.0   20303.920406   46472.889982    0.000000   
FLG_HAS_NOMINEE           20000.0       0.901150       0.298468    0.000000   
FLG_HAS_OLD_LOAN          20000.0       0.492950       0.499963    0.000000   
random                    20000.0       0.501933       0.290762    0.000011   

                                    25%            50%           75%  \
TARGET                         0.000000       0.000000  0.000000e+00   
AGE                           30.000000      38.000000  4.600000e+01   
BALANCE                    64754.025000  231675.845000  6.538768e+05   
SCR                          227.000000     364.000000  6.440000e+02   
HOLDING_PERIOD                 7.000000      15.000000  2.200000e+01   
LEN_OF_RLTN_IN_MNTH           79.000000     125.000000  1.720000e+02   
NO_OF_L_CR_TXNS                6.000000      10.000000  1.400000e+01   
NO_OF_L_DR_TXNS                2.000000       5.000000  7.000000e+00   
TOT_NO_OF_L_TXNS               9.000000      14.000000  2.100000e+01   
NO_OF_BR_CSH_WDL_DR_TXNS       1.000000       1.000000  2.000000e+00   
NO_OF_ATM_DR_TXNS              0.000000       1.000000  1.000000e+00   
NO_OF_NET_DR_TXNS              0.000000       0.000000  1.000000e+00   
NO_OF_MOB_DR_TXNS              0.000000       0.000000  0.000000e+00   
NO_OF_CHQ_DR_TXNS              0.000000       2.000000  4.000000e+00   
FLG_HAS_CC                     0.000000       0.000000  1.000000e+00   
AMT_ATM_DR                     0.000000    6900.000000  1.580000e+04   
AMT_BR_CSH_WDL_DR           2990.00000

In [10]:
data.skew()

TARGET                       2.259688
AGE                          0.092500
BALANCE                      2.777284
SCR                          0.555697
HOLDING_PERIOD               0.071600
LEN_OF_RLTN_IN_MNTH         -0.004770
NO_OF_L_CR_TXNS              2.462778
NO_OF_L_DR_TXNS              3.917768
TOT_NO_OF_L_TXNS             2.809076
NO_OF_BR_CSH_WDL_DR_TXNS     2.339490
NO_OF_ATM_DR_TXNS            8.469950
NO_OF_NET_DR_TXNS            4.244843
NO_OF_MOB_DR_TXNS           11.017740
NO_OF_CHQ_DR_TXNS            1.563019
FLG_HAS_CC                   0.845091
AMT_ATM_DR                   3.653754
AMT_BR_CSH_WDL_DR            0.308280
AMT_CHQ_DR                   7.184517
AMT_NET_DR                   0.998343
AMT_MOB_DR                   2.163102
AMT_L_DR                     2.528302
FLG_HAS_ANY_CHGS             2.483318
AMT_OTH_BK_ATM_USG_CHGS     14.074816
AMT_MIN_BAL_NMC_CHGS        11.340454
NO_OF_IW_CHQ_BNC_TXNS        4.574691
NO_OF_OW_CHQ_BNC_TXNS        4.441111
AVG_AMT_PER_

**Observations-**
* The set is generally right-skewed, other than a few left-skewed columns like FLG_HAS_NOMINEE.
* Some of the columns have right-skewedness in the double digits, such as NO_OF_MOB_DR_TXNS, AMT_OTH_BK_ATM_USG_CHGS, and AMT_MIN_BAL_NMC_CHGS.
* There may exists outliers that need treatment.

In [11]:
#Drop CUST_ID since it is essentially synonymous with the row and random as it is a randomly assigned number.
data.drop(['CUST_ID','random'],axis=1,inplace=True)

In [12]:
#Check our data-types again before EDA
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20000 entries, 0 to 19999
Data columns (total 38 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   TARGET                    20000 non-null  int64  
 1   AGE                       20000 non-null  int64  
 2   GENDER                    20000 non-null  object 
 3   BALANCE                   20000 non-null  float64
 4   OCCUPATION                20000 non-null  object 
 5   AGE_BKT                   20000 non-null  object 
 6   SCR                       20000 non-null  int64  
 7   HOLDING_PERIOD            20000 non-null  int64  
 8   ACC_TYPE                  20000 non-null  object 
 9   ACC_OP_DATE               20000 non-null  object 
 10  LEN_OF_RLTN_IN_MNTH       20000 non-null  int64  
 11  NO_OF_L_CR_TXNS           20000 non-null  int64  
 12  NO_OF_L_DR_TXNS           20000 non-null  int64  
 13  TOT_NO_OF_L_TXNS          20000 non-null  int64  
 14  NO_OF_

# **EDA**

### Univariate analysis

In [13]:
# function to plot a boxplot and a histogram along the same scale.


def histogram_boxplot(data, feature, figsize=(12, 7), kde=False, bins=None):
    """
    Boxplot and histogram combined

    data: dataframe
    feature: dataframe column
    figsize: size of figure (default (12,7))
    kde: whether to show the density curve (default False)
    bins: number of bins for histogram (default None)
    """
    f2, (ax_box2, ax_hist2) = plt.subplots(
        nrows=2,  # Number of rows of the subplot grid= 2
        sharex=True,  # x-axis will be shared among all subplots
        gridspec_kw={"height_ratios": (0.25, 0.75)},
        figsize=figsize,
    )  # creating the 2 subplots
    sns.boxplot(
        data=data, x=feature, ax=ax_box2, showmeans=True, color="violet"
    )  # boxplot will be created and a star will indicate the mean value of the column
    sns.histplot(
        data=data, x=feature, kde=kde, ax=ax_hist2, bins=bins, palette="winter"
    ) if bins else sns.histplot(
        data=data, x=feature, kde=kde, ax=ax_hist2
    )  # For histogram
    ax_hist2.axvline(
        data[feature].mean(), color="green", linestyle="--"
    )  # Add mean to the histogram
    ax_hist2.axvline(
        data[feature].median(), color="black", linestyle="-"
    )  # Add median to the histogram

### Observations on TARGET

In [14]:
histogram_boxplot(data,'TARGET')

* The distribution is right-skewed and binary, most did not respond.
* Outliers to the right.

### Observations on AGE

In [15]:
histogram_boxplot(data,'AGE')

* The distribution is fairly symmetric, but right-skewed.

### Observations on BALANCE

In [16]:
histogram_boxplot(data,'BALANCE')

* The distribution is right-skewed, there are also many outliers to the right.

### Observations on SCR

In [17]:
histogram_boxplot(data,'SCR')

* The distribution is right-skewed.

### Observations on HOLDING_PERIOD

In [18]:
histogram_boxplot(data,'HOLDING_PERIOD')

* The distribution is fairly symmetric, but slightly right-skewed.

### Observations on LEN_OF_RLTN_IN_MNTH

In [19]:
histogram_boxplot(data,'LEN_OF_RLTN_IN_MNTH')

* The distribution is fairly symmetric, but slight left-skewed.

### Observations on NO_OF_L_CR_TXNS

In [20]:
histogram_boxplot(data,'NO_OF_L_CR_TXNS')

* The distribution is right-skewed and there are many outliers to the right.

### Observations on NO_OF_L_DR_TXNS

In [21]:
histogram_boxplot(data,'NO_OF_L_DR_TXNS')

* The distribution is slightly right-skewed.
* Has many outliers to the right.

### Observations on TOT_NO_OF_L_TXNS

In [22]:
histogram_boxplot(data,'TOT_NO_OF_L_TXNS')

* The distribution is right-skewed and has many outliers to the right.

### Observations on NO_OF_BR_CSH_WDL_DR_TXNS

In [23]:
histogram_boxplot(data,'NO_OF_BR_CSH_WDL_DR_TXNS')

* The distribution is right-skewed and has many outliers to the right.

### Observations on NO_OF_ATM_DR_TXNS

In [24]:
histogram_boxplot(data,'NO_OF_ATM_DR_TXNS')

* The distribution is slightly right-skewed and has many outliers to the right.

### Observations on NO_OF_NET_DR_TXNS

In [25]:
histogram_boxplot(data,'NO_OF_NET_DR_TXNS')

* This distribution is right-skewed and has outliers to the right.

### Observations on NO_OF_MOB_DR_TXNS

In [26]:
histogram_boxplot(data,'NO_OF_MOB_DR_TXNS')

* This distribution is right-skewed and has outliers to the right.

### Observations on FLG_HAS_CC

In [27]:
histogram_boxplot(data,'FLG_HAS_CC')

* This distribution is right-skewed.

### Observations on AMT_ATM_DR

In [28]:
histogram_boxplot(data,'AMT_ATM_DR')

* This distribution is right-skewed and has many outliers to the right.

### Observations on AMT_BR_CSH_WDL_DR

In [29]:
histogram_boxplot(data,'AMT_BR_CSH_WDL_DR')

* This distribution is right-skewed.

### Observations on AMT_CHQ_DR

In [30]:
histogram_boxplot(data,'AMT_CHQ_DR')

* This distribution is right-skewed and has many outliers to the right.

### Observations on AMT_NET_DR

In [31]:
histogram_boxplot(data,'AMT_NET_DR')

* This distribution is right-skewed.

### Observations on AMT_L_DR

In [32]:
histogram_boxplot(data,'AMT_L_DR')

* This distribution is right-skewed and has outliers to the right.


### Observations on FLG_HAS_ANY_CHGS

In [33]:
histogram_boxplot(data,'FLG_HAS_ANY_CHGS')

* This binary distribution is right-skewed and has outliers to the right.

### Observations on AMT_OTH_BK_ATM_USG_CHGS

In [34]:
histogram_boxplot(data,'AMT_OTH_BK_ATM_USG_CHGS')

* This distribution is right-skewed and has outliers to the right.

### Observations on AMT_MIN_BAL_NMC_CHGS

In [35]:
histogram_boxplot(data,'AMT_MIN_BAL_NMC_CHGS')

* This distribution is right-skewed and has an outlier to the right.

### Observations on NO_OF_IW_CHQ_BNC_TXNS

In [36]:
histogram_boxplot(data,'NO_OF_IW_CHQ_BNC_TXNS')

* This distribution is right-skewed and has outliers to the right.

### Observations on NO_OF_OW_CHQ_BNC_TXNS

In [37]:
histogram_boxplot(data,'NO_OF_OW_CHQ_BNC_TXNS')

* This distribution is right-skewed and has outliers to the right.

### Observations on AVG_AMT_PER_ATM_TXN

In [38]:
histogram_boxplot(data,'AVG_AMT_PER_ATM_TXN')

* This distribution is right-skewed.

### Observations on AVG_AMT_PER_CSH_WDL_TXN

In [39]:
histogram_boxplot(data,'AVG_AMT_PER_CSH_WDL_TXN')

* This distribution is right-skewed and has outliers to the right.

### Observations on AVG_AMT_PER_CHQ_TXN 

In [40]:
histogram_boxplot(data,'AVG_AMT_PER_CHQ_TXN')

* This distribution is right-skewed and has outliers to the right.

### Observations on AVG_AMT_PER_NET_TXN

In [41]:
histogram_boxplot(data,'AVG_AMT_PER_NET_TXN')

* This distribution is right-skewed and has outliers to the right.

### Observations on AVG_AMT_PER_MOB_TXN

In [42]:
histogram_boxplot(data,'AVG_AMT_PER_MOB_TXN')

* This distribution is right-skewed and has outliers to the right.

### Observations on FLG_HAS_NOMINEE 

In [43]:
histogram_boxplot(data,'FLG_HAS_NOMINEE')

* This binary distribution is left-skewed and has outliers to the left.

### Observations on FLG_HAS_OLD_LOAN

In [44]:
histogram_boxplot(data,'FLG_HAS_OLD_LOAN')

* This binary distribution is right-skewed.

In [45]:
# function to create labeled barplots


def labeled_barplot(data, feature, perc=False, n=None):
    """
    Barplot with percentage at the top

    data: dataframe
    feature: dataframe column
    perc: whether to display percentages instead of count (default is False)
    n: displays the top n category levels (default is None, i.e., display all levels)
    """

    total = len(data[feature])  # length of the column
    count = data[feature].nunique()
    if n is None:
        plt.figure(figsize=(count + 1, 5))
    else:
        plt.figure(figsize=(n + 1, 5))

    plt.xticks(rotation=90, fontsize=15)
    ax = sns.countplot(
        data=data,
        x=feature,
        palette="Paired",
        order=data[feature].value_counts().index[:n].sort_values(),
    )

    for p in ax.patches:
        if perc == True:
            label = "{:.1f}%".format(
                100 * p.get_height() / total
            )  # percentage of each class of the category
        else:
            label = p.get_height()  # count of each level of the category

        x = p.get_x() + p.get_width() / 2  # width of the plot
        y = p.get_height()  # height of the plot

        ax.annotate(
            label,
            (x, y),
            ha="center",
            va="center",
            size=12,
            xytext=(0, 5),
            textcoords="offset points",
        )  # annotate the percentage

    plt.show()  # show the plot

### Observations on GENDER

In [46]:
labeled_barplot(data,"GENDER",perc=True)

* 71.9% of the customers are males.
* 27.2% are females.
* 1.0% are "other."

In [47]:
#Convert to categorical and check dtypes
data["GENDER"] = data["GENDER"].astype("category")
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20000 entries, 0 to 19999
Data columns (total 38 columns):
 #   Column                    Non-Null Count  Dtype   
---  ------                    --------------  -----   
 0   TARGET                    20000 non-null  int64   
 1   AGE                       20000 non-null  int64   
 2   GENDER                    20000 non-null  category
 3   BALANCE                   20000 non-null  float64 
 4   OCCUPATION                20000 non-null  object  
 5   AGE_BKT                   20000 non-null  object  
 6   SCR                       20000 non-null  int64   
 7   HOLDING_PERIOD            20000 non-null  int64   
 8   ACC_TYPE                  20000 non-null  object  
 9   ACC_OP_DATE               20000 non-null  object  
 10  LEN_OF_RLTN_IN_MNTH       20000 non-null  int64   
 11  NO_OF_L_CR_TXNS           20000 non-null  int64   
 12  NO_OF_L_DR_TXNS           20000 non-null  int64   
 13  TOT_NO_OF_L_TXNS          20000 non-null  int6

### Observations on OCCUPATION

In [48]:
labeled_barplot(data,"OCCUPATION",perc=True)

* Fairly evenly distrubted amongst the 4 categories, the dominant category is SAL with 29.3% and the least dominant category is Self-EMP with 17.8% of the set.
* Both types of self-employment may need to be merged.

In [49]:
#Convert to categorical and check dtypes
data["OCCUPATION"] = data["OCCUPATION"].astype("category")
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20000 entries, 0 to 19999
Data columns (total 38 columns):
 #   Column                    Non-Null Count  Dtype   
---  ------                    --------------  -----   
 0   TARGET                    20000 non-null  int64   
 1   AGE                       20000 non-null  int64   
 2   GENDER                    20000 non-null  category
 3   BALANCE                   20000 non-null  float64 
 4   OCCUPATION                20000 non-null  category
 5   AGE_BKT                   20000 non-null  object  
 6   SCR                       20000 non-null  int64   
 7   HOLDING_PERIOD            20000 non-null  int64   
 8   ACC_TYPE                  20000 non-null  object  
 9   ACC_OP_DATE               20000 non-null  object  
 10  LEN_OF_RLTN_IN_MNTH       20000 non-null  int64   
 11  NO_OF_L_CR_TXNS           20000 non-null  int64   
 12  NO_OF_L_DR_TXNS           20000 non-null  int64   
 13  TOT_NO_OF_L_TXNS          20000 non-null  int6

### Observations on AGE_BKT

In [50]:
labeled_barplot(data,"AGE_BKT",perc=True)

* This column is already binned to 7 classes, the dominant classes are 26-30 and 31-35.

In [51]:
#Convert to categorical and check dtypes
data["AGE_BKT"] = data["AGE_BKT"].astype("category")
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20000 entries, 0 to 19999
Data columns (total 38 columns):
 #   Column                    Non-Null Count  Dtype   
---  ------                    --------------  -----   
 0   TARGET                    20000 non-null  int64   
 1   AGE                       20000 non-null  int64   
 2   GENDER                    20000 non-null  category
 3   BALANCE                   20000 non-null  float64 
 4   OCCUPATION                20000 non-null  category
 5   AGE_BKT                   20000 non-null  category
 6   SCR                       20000 non-null  int64   
 7   HOLDING_PERIOD            20000 non-null  int64   
 8   ACC_TYPE                  20000 non-null  object  
 9   ACC_OP_DATE               20000 non-null  object  
 10  LEN_OF_RLTN_IN_MNTH       20000 non-null  int64   
 11  NO_OF_L_CR_TXNS           20000 non-null  int64   
 12  NO_OF_L_DR_TXNS           20000 non-null  int64   
 13  TOT_NO_OF_L_TXNS          20000 non-null  int6

### Observations on ACC_TYPE

In [52]:
labeled_barplot(data,"ACC_TYPE",perc=True)

* Binary distribution and a majority of accounts are savings accounts with 78.8% of the set.

In [53]:
#Convert to categorical and check dtypes
data["ACC_TYPE"] = data["ACC_TYPE"].astype("category")
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20000 entries, 0 to 19999
Data columns (total 38 columns):
 #   Column                    Non-Null Count  Dtype   
---  ------                    --------------  -----   
 0   TARGET                    20000 non-null  int64   
 1   AGE                       20000 non-null  int64   
 2   GENDER                    20000 non-null  category
 3   BALANCE                   20000 non-null  float64 
 4   OCCUPATION                20000 non-null  category
 5   AGE_BKT                   20000 non-null  category
 6   SCR                       20000 non-null  int64   
 7   HOLDING_PERIOD            20000 non-null  int64   
 8   ACC_TYPE                  20000 non-null  category
 9   ACC_OP_DATE               20000 non-null  object  
 10  LEN_OF_RLTN_IN_MNTH       20000 non-null  int64   
 11  NO_OF_L_CR_TXNS           20000 non-null  int64   
 12  NO_OF_L_DR_TXNS           20000 non-null  int64   
 13  TOT_NO_OF_L_TXNS          20000 non-null  int6

### Observations on ACC_OP_DATE

In [54]:
#Convert to date/time format and check dtypes
data["ACC_OP_DATE"] = data["ACC_OP_DATE"].astype("datetime64")
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20000 entries, 0 to 19999
Data columns (total 38 columns):
 #   Column                    Non-Null Count  Dtype         
---  ------                    --------------  -----         
 0   TARGET                    20000 non-null  int64         
 1   AGE                       20000 non-null  int64         
 2   GENDER                    20000 non-null  category      
 3   BALANCE                   20000 non-null  float64       
 4   OCCUPATION                20000 non-null  category      
 5   AGE_BKT                   20000 non-null  category      
 6   SCR                       20000 non-null  int64         
 7   HOLDING_PERIOD            20000 non-null  int64         
 8   ACC_TYPE                  20000 non-null  category      
 9   ACC_OP_DATE               20000 non-null  datetime64[ns]
 10  LEN_OF_RLTN_IN_MNTH       20000 non-null  int64         
 11  NO_OF_L_CR_TXNS           20000 non-null  int64         
 12  NO_OF_L_DR_TXNS   

All potential categorical data columns have been checked and converted to the proper dtype.

## Bivariate Analysis

In [55]:
plt.figure(figsize=(20,14))
sns.heatmap(data.corr(),annot=True,vmin=-1,vmax=1,fmt='.1g',cmap="Spectral")
plt.show()

* There appears to be multiple columns with strong positive correlations to each other, some columns with strong negative correlations to each other, and some that are even correlated 1:1 with other columns and can likely be dropped from the set.

In [56]:
data.corr()

TARGET       AGE   BALANCE       SCR  \
TARGET                    1.000000  0.022977 -0.067062  0.078909   
AGE                       0.022977  1.000000 -0.122587 -0.005871   
BALANCE                  -0.067062 -0.122587  1.000000 -0.065300   
SCR                       0.078909 -0.005871 -0.065300  1.000000   
HOLDING_PERIOD           -0.135447 -0.012060  0.039241 -0.025174   
LEN_OF_RLTN_IN_MNTH      -0.023618 -0.008279 -0.000855  0.009034   
NO_OF_L_CR_TXNS           0.120284  0.030170 -0.047652 -0.013179   
NO_OF_L_DR_TXNS           0.122954  0.019602 -0.043431  0.009389   
TOT_NO_OF_L_TXNS          0.134504  0.028932 -0.051299 -0.004952   
NO_OF_BR_CSH_WDL_DR_TXNS  0.101909  0.024266 -0.034814  0.004827   
NO_OF_ATM_DR_TXNS         0.082359  0.006757 -0.030579  0.014910   
NO_OF_NET_DR_TXNS         0.097071  0.012268 -0.037018  0.006476   
NO_OF_MOB_DR_TXNS         0.054477  0.001957 -0.021012  0.001236   
NO_OF_CHQ_DR_TXNS         0.103531  0.021800 -0.032516  0.008708   
FLG_HAS_CC                0.100513  0.014168 -0.019638  0.030116   
AMT_ATM_DR                0.083251  0.019702 -0.035987  0.009744   
AMT_BR_CSH_WDL_DR         0.048360  0.010351 -0.020901 -0.019841   
AMT_CHQ_DR                0.060796  0.012189 -0.012511  0.004519   
AMT_NET_DR                0.062121  0.004390 -0.029335  0.023654   
AMT_MOB_DR                0.083215  0.028908 -0.026694 -0.012588   
AMT_L_DR                  0.095351  0.016725 -0.033193  0.003485   
FLG_HAS_ANY_CHGS          0.045783 -0.008505 -0.006826  0.014432   
AMT_OTH_BK_ATM_USG_CHGS   0.028417 -0.007980 -0.010949  0.007902   
AMT_MIN_BAL_NMC_CHGS      0.005053  0.016437 -0.061819  0.008682   
NO_OF_IW_CHQ_BNC_TXNS     0.011601 -0.014412 -0.009934  0.000864   
NO_OF_OW_CHQ_BNC_TXNS     0.027411 -0.005830 -0.014364  0.006541   
AVG_AMT_PER_ATM_TXN       0.074995  0.018692 -0.035360  0.019440   
AVG_AMT_PER_CSH_WDL_TXN  -0.002157 -0.005078  0.005807 -0.022531   
AVG_AMT_PER_CHQ_TXN       0.065239  0.014456 -0.023732  0.008267   
AVG_AMT_PER_NET_TXN       0.025707  0.000079 -0.019149  0.025731   
AVG_AMT_PER_MOB_TXN       0.076258  0.028288 -0.028727 -0.015246   
FLG_HAS_NOMINEE           0.011784  0.016856 -0.004205 -0.001952   
FLG_HAS_OLD_LOAN         -0.011555  0.010269 -0.028707  0.013333   

                          HOLDING_PERIOD  LEN_OF_RLTN_IN_MNTH  \
TARGET                         -0.135447            -0.023618   
AGE                            -0.012060            -0.008279   
BALANCE                         0.039241            -0.000855   
SCR                            -0.025174             0.009034   
HOLDING_PERIOD                  1.000000             0.011448   
LEN_OF_RLTN_IN_MNTH             0.011448             1.000000   
NO_OF_L_CR_TXNS                -0.008474             0.021719   
NO_OF_L_DR_TXNS                -0.347855             0.001878   
TOT_NO_OF_L_TXNS               -0.156350             0.015777   
NO_OF_BR_CSH_WDL_DR_TXNS       -0.196455            -0.003073   
NO_OF_ATM_DR_TXNS              -0.181790            -0.000047   
NO_OF_NET_DR_TXNS              -0.250650             0.005921   
NO_OF_MOB_DR_TXNS              -0.109956            -0.003159   
NO_OF_CHQ_DR_TXNS              -0.473639             0.005685   
FLG_HAS_CC                      0.001694            -0.015797   
AMT_ATM_DR                     -0.173627            -0.010014   
AMT_BR_CSH_WDL_DR              -0.048038            -0.001699   
AMT_CHQ_DR                     -0.173118             0.013796   
AMT_NET_DR                     -0.198319             0.001536   
AMT_MOB_DR                     -0.104097             0.004806   
AMT_L_DR                       -0.228821             0.008522   
FLG_HAS_ANY_CHGS               -0.006915             0.019448   
AMT_OTH_BK_ATM_USG_CHGS        -0.077078            -0.006432   
AMT_MIN_BAL_NMC_CHGS           -0.006945            -0.008239   
NO_OF_IW_CHQ_BNC_TXNS          -0.040950             0.001705   
NO_OF_OW_CHQ_BNC_TXNS    

* Let's examine some of the categories/flags distributed about the target field.
* We would like to know which columns may have higher responding rates.

In [57]:
sns.barplot(x="OCCUPATION", y="TARGET", data=data,)

* All categories of occupation have similar distribution widths, but self-empoyed people generally have higher response rates than the others.

In [58]:
sns.barplot(x="GENDER", y="TARGET", data=data,)

* The other category of gender has the highest response rate relative to the other genders, it is also significantly smaller than the other genders in this set though.

In [59]:
sns.barplot(x="AGE_BKT", y="TARGET", data=data,)

* We see the highest response rates in the 41-45 year old bin, but it is fairly symmetric across the age bins.

In [60]:
sns.barplot(x="ACC_TYPE", y="TARGET", data=data,)

* Current accounts have a higher response rate than savings accounts.

In [61]:
sns.barplot(x="FLG_HAS_CC", y="TARGET", data=data,)

* We see a higher response rate among those that have a card.

In [62]:
sns.barplot(x="FLG_HAS_ANY_CHGS", y="TARGET", data=data,)

* Similarly we see a higher response rate among those that have banking charges.

In [63]:
sns.barplot(x="FLG_HAS_NOMINEE", y="TARGET", data=data,)

* We see a slightly higher response rate from those that have a nominee.

In [64]:
sns.barplot(x="FLG_HAS_OLD_LOAN", y="TARGET", data=data,)

* We see a slightly higher response rate among those that do not have an old loan.

#Summary of EDA

* Target: Right skewed binary distribution, most are responders.
* Age: Slightly right-skewed, fairly symmetric distribution of ages.
* Gender: Predominantly male set.
* Balance: Skewed right about a median of $231K, many outliers to the right.
* Occupation: Fairly evenly distributed but dominated by salaried employees.
* Age_BKT: Majority of the set is 26-45, primarily in the 26-30 bin.
* SCR: Slightly right skewed.
* HOLDING_PERIOD: Fairly symmetric with a large portion of the set having a low holding period.
* ACC_TYPE: Predominantly savings accounts
* LEN_OF_RLTN_IN_MNTH: Fairly symmetric distribution, varying lengths of membership across the set.
* NO_OF_L_CR_TXNS: Right skewed with many right outliers.
* NO_OF_L_DR_TXNS: Right skewed with many right outliers.
* TOT_NO_OF_L_TXNS: Right skewed with many right outliers.
* NO_OF_BR_CSH_WDL_DR_TXNS: Right skewed with some outliers to the right.
* NO_OF_ATM_DR_TXNS: Many outliers to the right.
* NO_OF_NET_DR_TXNS: Many outliers to the right.
* NO_OF_MOB_DR_TXNS: Right skewed with outliers to the right.
* NO_OF_CHQ_DR_TXNS: Right skewed.
* FLG_HAS_CC: Right skewed, most do not have a credit card.
* AMT_ATM_DR: Right skewed with many outliers to the right.
* AMT_BR_CSH_WDL_DR: Right skewed, large portion of the set has very low value.
* AMT_CHQ_DR: Right skewed with many outliers to the right.
* AMT_NET_DR: Right skewed.
* AMT_MOB_DR: Right skewed
* AMT_L_DR: Right skewed and many outliers.
* FLG_HAS_ANY_CHGS: Right skewed, most of the set does not.
* AMT_OTH_BK_ATM_USG_CHGS: Right skewed with right outliers.
* AMT_MIN_BAL_NMC_CHGS: Right skewed with an outlier.
* NO_OF_IW_CHQ_BNC_TXNS: Right skewed and outliers.
* NO_OF_OW_CHQ_BNC_TXNS: Right skewed with similar outliers to IW.
* AVG_AMT_PER_ATM_TXN: Right skewed, large portion of set has very small value here.
* AVG_AMT_PER_CSH_WDL_TXN: Right skewed and many outliers right past the IQR.
* AVG_AMT_PER_CHQ_TXN: Many outliers to the right and rightly skewed distribution.
* AVG_AMT_PER_NET_TXN: Right skewed and outliers.
* AVG_AMT_PER_MOB_TXN: Rightly skewed distribution.
* FLG_HAS_NOMINEE: Skewed left, predominantly no nominee.
* FLG_HAS_OLD_LOAN: Flag: Fairly symmetric distribution.

#Data Pre-processing



##Outlier Detection

### Let's find the percentage of outliers, in each column of the data, using IQR.

In [65]:
Q1 = data.quantile(0.25)  # To find the 25th percentile and 75th percentile.
Q3 = data.quantile(0.75)

IQR = Q3 - Q1  # Inter Quantile Range (75th perentile - 25th percentile)

lower = (
    Q1 - 1.5 * IQR
)  # Finding lower and upper bounds for all values. All values outside these bounds are outliers
upper = Q3 + 1.5 * IQR

In [66]:
(
    (data.select_dtypes(include=["float64", "int64"]) < lower)
    | (data.select_dtypes(include=["float64", "int64"]) > upper)
).sum() / len(data) * 100

TARGET                      12.560
AGE                          0.000
BALANCE                      8.160
SCR                          0.000
HOLDING_PERIOD               0.000
LEN_OF_RLTN_IN_MNTH          0.000
NO_OF_L_CR_TXNS              8.960
NO_OF_L_DR_TXNS             11.745
TOT_NO_OF_L_TXNS             9.560
NO_OF_BR_CSH_WDL_DR_TXNS    13.250
NO_OF_ATM_DR_TXNS            4.600
NO_OF_NET_DR_TXNS           13.765
NO_OF_MOB_DR_TXNS           22.745
NO_OF_CHQ_DR_TXNS            1.160
FLG_HAS_CC                   0.000
AMT_ATM_DR                   5.590
AMT_BR_CSH_WDL_DR            0.000
AMT_CHQ_DR                   9.790
AMT_NET_DR                   0.000
AMT_MOB_DR                  22.745
AMT_L_DR                     1.955
FLG_HAS_ANY_CHGS            11.060
AMT_OTH_BK_ATM_USG_CHGS      0.705
AMT_MIN_BAL_NMC_CHGS         0.760
NO_OF_IW_CHQ_BNC_TXNS        4.260
NO_OF_OW_CHQ_BNC_TXNS        4.435
AVG_AMT_PER_ATM_TXN          0.000
AVG_AMT_PER_CSH_WDL_TXN      1.230
AVG_AMT_PER_CHQ_TXN 

###We will not be treating for the detected outliers in this set, despite some columns having double digit percentages of said outliers. Some outliers have been identified earlier in the EDA. I do not feel that removing the outliers from any of these columns is ncessary and we would like to include these outliers in our model to account for these outliers' effects on the trends of our customers.

##No null or missing values in our set.

In [67]:
data.isna().sum()

TARGET                      0
AGE                         0
GENDER                      0
BALANCE                     0
OCCUPATION                  0
AGE_BKT                     0
SCR                         0
HOLDING_PERIOD              0
ACC_TYPE                    0
ACC_OP_DATE                 0
LEN_OF_RLTN_IN_MNTH         0
NO_OF_L_CR_TXNS             0
NO_OF_L_DR_TXNS             0
TOT_NO_OF_L_TXNS            0
NO_OF_BR_CSH_WDL_DR_TXNS    0
NO_OF_ATM_DR_TXNS           0
NO_OF_NET_DR_TXNS           0
NO_OF_MOB_DR_TXNS           0
NO_OF_CHQ_DR_TXNS           0
FLG_HAS_CC                  0
AMT_ATM_DR                  0
AMT_BR_CSH_WDL_DR           0
AMT_CHQ_DR                  0
AMT_NET_DR                  0
AMT_MOB_DR                  0
AMT_L_DR                    0
FLG_HAS_ANY_CHGS            0
AMT_OTH_BK_ATM_USG_CHGS     0
AMT_MIN_BAL_NMC_CHGS        0
NO_OF_IW_CHQ_BNC_TXNS       0
NO_OF_OW_CHQ_BNC_TXNS       0
AVG_AMT_PER_ATM_TXN         0
AVG_AMT_PER_CSH_WDL_TXN     0
AVG_AMT_PE

## Data Preparation

In [68]:
# Dropping averages due to high correlations with resepective amounts, AGE since it is already binned,
# opening date since it is somewhat redundant with length, and the TARGET variable
X = data.drop(["TARGET", "AGE","ACC_OP_DATE","AVG_AMT_PER_ATM_TXN",
               "AVG_AMT_PER_CSH_WDL_TXN","AVG_AMT_PER_CHQ_TXN",
               "AVG_AMT_PER_NET_TXN","AVG_AMT_PER_MOB_TXN"],
               axis=1)

Y = data["TARGET"]
# Dummying all remaining categorical variables, other than the binary categories, and our dates.
# ACC_TYPE is binary, but dummying the two categories will get us the same result in our model.
X = pd.get_dummies(X, columns=["OCCUPATION","GENDER","ACC_TYPE",
                               "AGE_BKT"], drop_first=True)

# Splitting data in train and test sets
X_train, X_test, y_train, y_test = train_test_split(
    X, Y, test_size=0.30, random_state=1
)

In [69]:
print("Shape of Training set : ", X_train.shape)
print("Shape of test set : ", X_test.shape)
print("Percentage of classes in training set:")
print(y_train.value_counts(normalize=True))
print("Percentage of classes in test set:")
print(y_test.value_counts(normalize=True))

Shape of Training set :  (14000, 38)
Shape of test set :  (6000, 38)
Percentage of classes in training set:
0    0.873357
1    0.126643
Name: TARGET, dtype: float64
Percentage of classes in test set:
0    0.876833
1    0.123167
Name: TARGET, dtype: float64


#Model building - Logistic Regression




In [70]:
# defining a function to compute different metrics to check performance of a classification model built using sklearn


def model_performance_classification_sklearn_with_threshold(
    model, predictors, target, threshold=0.5
):
    """
    Function to compute different metrics, based on the threshold specified, to check classification model performance

    model: classifier
    predictors: independent variables
    target: dependent variable
    threshold: threshold for classifying the observation as class 1
    """

    # predicting using the independent variables
    pred_prob = model.predict_proba(predictors)[:, 1]
    pred_thres = pred_prob > threshold
    pred = np.round(pred_thres)

    acc = accuracy_score(target, pred)  # to compute Accuracy
    recall = recall_score(target, pred)  # to compute Recall
    precision = precision_score(target, pred)  # to compute Precision
    f1 = f1_score(target, pred)  # to compute F1-score

    # creating a dataframe of metrics
    df_perf = pd.DataFrame(
        {"Accuracy": acc, "Recall": recall, "Precision": precision, "F1": f1,},
        index=[0],
    )

    return df_perf

In [71]:
# defining a function to plot the confusion_matrix of a classification model built using sklearn
def confusion_matrix_sklearn_with_threshold(model, predictors, target, threshold=0.5):
    """
    To plot the confusion_matrix, based on the threshold specified, with percentages

    model: classifier
    predictors: independent variables
    target: dependent variable
    threshold: threshold for classifying the observation as class 1
    """
    pred_prob = model.predict_proba(predictors)[:, 1]
    pred_thres = pred_prob > threshold
    y_pred = np.round(pred_thres)

    cm = confusion_matrix(target, y_pred)
    labels = np.asarray(
        [
            ["{0:0.0f}".format(item) + "\n{0:.2%}".format(item / cm.flatten().sum())]
            for item in cm.flatten()
        ]
    ).reshape(2, 2)

    plt.figure(figsize=(6, 4))
    sns.heatmap(cm, annot=labels, fmt="")
    plt.ylabel("True label")
    plt.xlabel("Predicted label")

# Logistic Regression

In [72]:
# There are different solvers available in Sklearn logistic regression
# The newton-cg solver is faster for high-dimensional data

lg = LogisticRegression(solver="newton-cg", random_state=1)
model = lg.fit(X_train, y_train)

### Finding the coefficients

In [73]:
log_odds = lg.coef_[0]
pd.options.display.float_format = "{:.5f}".format
pd.DataFrame(log_odds, X_train.columns, columns=["coef"]).T

BALANCE     SCR  HOLDING_PERIOD  LEN_OF_RLTN_IN_MNTH  NO_OF_L_CR_TXNS  \
coef -0.00000 0.00003        -0.06678             -0.00616          0.01581   

      NO_OF_L_DR_TXNS  TOT_NO_OF_L_TXNS  NO_OF_BR_CSH_WDL_DR_TXNS  \
coef         -0.00470           0.00085                   0.01505   

      NO_OF_ATM_DR_TXNS  NO_OF_NET_DR_TXNS  NO_OF_MOB_DR_TXNS  \
coef            0.00414            0.00957            0.00596   

      NO_OF_CHQ_DR_TXNS  FLG_HAS_CC  AMT_ATM_DR  AMT_BR_CSH_WDL_DR  \
coef           -0.03942     0.02087    -0.00000           -0.00000   

      AMT_CHQ_DR  AMT_NET_DR  AMT_MOB_DR  AMT_L_DR  FLG_HAS_ANY_CHGS  \
coef     0.00000    -0.00000     0.00000  -0.00000           0.00272   

      AMT_OTH_BK_ATM_USG_CHGS  AMT_MIN_BAL_NMC_CHGS  NO_OF_IW_CHQ_BNC_TXNS  \
coef                 -0.00393              -0.00119               -0.00047   

      NO_OF_OW_CHQ_BNC_TXNS  FLG_HAS_NOMINEE  FLG_HAS_OLD_LOAN  \
coef                0.00006         -0.01793          -0.01551   

      OCCUPATION_SAL  OCCUPATION_SELF-EMP  OCCUPATION_SENP  GENDER_M  \
coef        -0.02274              0.01963         -0.01053  -0.00876   

      GENDER_O  ACC_TYPE_SA  AGE_BKT_31-35  AGE_BKT_36-40  AGE_BKT_41-45  \
coef  -0.00034     -0.02233       -0.00419       -0.00205        0.00696   

      AGE_BKT_46-50  AGE_BKT_<25  AGE_BKT_>50  
coef       -0.00072     -0.00471     -0.01110

### Coefficient interpretation

* The largest positive correlation exists with those that are self-employed with 0.01963. An increase in this and all other columns with a positive coefficient should lead to an an increase in responses.

* The largest negative coeffient exists for those that are salaried and those with savings accounts with -0.02274 and -0.02233, respectively. All columns with negative coefficients should lead to a decrease in response rate when they increase.

#### Checking model performance on the training set

In [74]:
# creating confusion matrix
confusion_matrix_sklearn_with_threshold(lg, X_train, y_train)

In [75]:
log_reg_model_train_perf = model_performance_classification_sklearn_with_threshold(
    lg, X_train, y_train
)

print("Training performance:")
log_reg_model_train_perf

Training performance:


Accuracy  Recall  Precision      F1
0   0.87350 0.00677    0.54545 0.01337

* High accuracy but a low recall and low f1 score.

#### ROC-AUC

In [76]:
logit_roc_auc_train = roc_auc_score(y_train, lg.predict_proba(X_train)[:, 1])
fpr, tpr, thresholds = roc_curve(y_train, lg.predict_proba(X_train)[:, 1])
plt.figure(figsize=(7, 5))
plt.plot(fpr, tpr, label="Logistic Regression (area = %0.2f)" % logit_roc_auc_train)
plt.plot([0, 1], [0, 1], "r--")
plt.xlim([0.0, 1.0])
plt.ylim([0.0, 1.05])
plt.xlabel("False Positive Rate")
plt.ylabel("True Positive Rate")
plt.title("Receiver operating characteristic")
plt.legend(loc="lower right")
plt.show()

* Logistic Regression model is giving decent performance with a score of 0.65

### Model Performance Improvement

### Optimal threshold using AUC-ROC curve

In [77]:
# Optimal threshold as per AUC-ROC curve
# The optimal cut off would be where tpr is high and fpr is low
fpr, tpr, thresholds = roc_curve(y_train, lg.predict_proba(X_train)[:, 1])

optimal_idx = np.argmax(tpr - fpr)
optimal_threshold_auc_roc = thresholds[optimal_idx]
print(optimal_threshold_auc_roc)

0.1369793779721916


* Optimal threshold determined to be ~0.137

#### Checking model performance on the training set

In [78]:
# creating confusion matrix
confusion_matrix_sklearn_with_threshold(
    lg, X_train, y_train, threshold=optimal_threshold_auc_roc
)

In [79]:
# checking model performance for this model
log_reg_model_train_perf_threshold_auc_roc = model_performance_classification_sklearn_with_threshold(
    lg, X_train, y_train, threshold=optimal_threshold_auc_roc
)
print("Training performance:")
log_reg_model_train_perf_threshold_auc_roc

Training performance:


Accuracy  Recall  Precision      F1
0   0.60714 0.61985    0.18549 0.28553

* Recall and f1 have significantly increased with the optimal threshold. Accuracy and precision have decreased, this model is significantly more balanced.

#### Let's use Precision-Recall curve and see if we can find a better threshold

In [80]:
y_scores = lg.predict_proba(X_train)[:, 1]
prec, rec, tre = precision_recall_curve(y_train, y_scores,)


def plot_prec_recall_vs_tresh(precisions, recalls, thresholds):
    plt.plot(thresholds, precisions[:-1], "b--", label="precision")
    plt.plot(thresholds, recalls[:-1], "g--", label="recall")
    plt.xlabel("Threshold")
    plt.legend(loc="upper left")
    plt.ylim([0, 1])


plt.figure(figsize=(10, 7))
plot_prec_recall_vs_tresh(prec, rec, tre)
plt.show()

* At threshold around 0.24 we get a higher recall and a good precision.

In [81]:
# setting the threshold
optimal_threshold_curve = 0.24

#### Checking model performance on training set

In [82]:
# creating confusion matrix
confusion_matrix_sklearn_with_threshold(
    lg, X_train, y_train, threshold=optimal_threshold_curve
)

In [83]:
log_reg_model_train_perf_threshold_curve = model_performance_classification_sklearn_with_threshold(
    lg, X_train, y_train, threshold=optimal_threshold_curve
)
print("Training performance:")
log_reg_model_train_perf_threshold_curve

Training performance:


Accuracy  Recall  Precision      F1
0   0.80871 0.25268    0.24875 0.25070

* Model is performing better on the training set overall than the first model.
* Model is well-balanced.

### Let's check the performance on the test set

**Using model with default threshold**

In [84]:
# creating confusion matrix
confusion_matrix_sklearn_with_threshold(lg, X_test, y_test)

In [85]:
log_reg_model_test_perf = model_performance_classification_sklearn_with_threshold(
    lg, X_test, y_test
)

print("Test performance:")
log_reg_model_test_perf

Test performance:


Accuracy  Recall  Precision      F1
0   0.87683 0.00947    0.50000 0.01859

* This model is not performing well at all with default threshold on the test set.

* ROC curve on test set

In [86]:
logit_roc_auc_test = roc_auc_score(y_test, lg.predict_proba(X_test)[:, 1])
fpr, tpr, thresholds = roc_curve(y_test, lg.predict_proba(X_test)[:, 1])
plt.figure(figsize=(7, 5))
plt.plot(fpr, tpr, label="Logistic Regression (area = %0.2f)" % logit_roc_auc_test)
plt.plot([0, 1], [0, 1], "r--")
plt.xlim([0.0, 1.0])
plt.ylim([0.0, 1.05])
plt.xlabel("False Positive Rate")
plt.ylabel("True Positive Rate")
plt.title("Receiver operating characteristic")
plt.legend(loc="lower right")
plt.show()

**Using model with threshold=0.13** 

In [87]:
# creating confusion matrix
confusion_matrix_sklearn_with_threshold(
    lg, X_test, y_test, threshold=optimal_threshold_auc_roc
)

In [88]:
# checking model performance for this model
log_reg_model_test_perf_threshold_auc_roc = model_performance_classification_sklearn_with_threshold(
    lg, X_test, y_test, threshold=optimal_threshold_auc_roc
)
print("Test performance:")
log_reg_model_test_perf_threshold_auc_roc

Test performance:


Accuracy  Recall  Precision      F1
0   0.60967 0.59811    0.17772 0.27402

* More balanced model than the default threshold on test set.

**Using model with threshold = 0.24**

In [89]:
# creating confusion matrix
confusion_matrix_sklearn_with_threshold(
    lg, X_test, y_test, threshold=optimal_threshold_curve
)

In [90]:
log_reg_model_test_perf_threshold_curve = model_performance_classification_sklearn_with_threshold(
    lg, X_test, y_test, threshold=optimal_threshold_curve
)
print("Test performance:")
log_reg_model_test_perf_threshold_curve

Test performance:


Accuracy  Recall  Precision      F1
0   0.80733 0.22733    0.22311 0.22520

* Also well-balanced with higher accuracy and precision than the 0.13 model; however, the f1 score and recall have decreased significantly.

### Model performance summary

In [91]:
# training performance comparison

models_train_comp_df = pd.concat(
    [
        log_reg_model_train_perf.T,
        log_reg_model_train_perf_threshold_auc_roc.T,
        log_reg_model_train_perf_threshold_curve.T,
    ],
    axis=1,
)
models_train_comp_df.columns = [
    "Logistic Regression sklearn",
    "Logistic Regression ROC-AUC",
    "Logistic Regression Precision-Recall",
]

print("Training performance comparison:")
models_train_comp_df

Training performance comparison:


Logistic Regression sklearn  Logistic Regression ROC-AUC  \
Accuracy                       0.87350                      0.60714   
Recall                         0.00677                      0.61985   
Precision                      0.54545                      0.18549   
F1                             0.01337                      0.28553   

           Logistic Regression Precision-Recall  
Accuracy                                0.80871  
Recall                                  0.25268  
Precision                               0.24875  
F1                                      0.25070

* Both the ROC-AUC optimized threshold and Precision-Recall optimized threshold models have well-balanced performance, but the Precision-Recall seems to have the best balance of performance.

In [92]:
# testing performance comparison

models_test_comp_df = pd.concat(
    [
        log_reg_model_test_perf.T,
        log_reg_model_test_perf_threshold_auc_roc.T,
        log_reg_model_test_perf_threshold_curve.T,
    ],
    axis=1,
)
models_test_comp_df.columns = [
    "Logistic Regression sklearn",
    "Logistic Regression ROC-AUC",
    "Logistic Regression Precision-Recall",
]

print("Test set performance comparison:")
models_test_comp_df

Test set performance comparison:


Logistic Regression sklearn  Logistic Regression ROC-AUC  \
Accuracy                       0.87683                      0.60967   
Recall                         0.00947                      0.59811   
Precision                      0.50000                      0.17772   
F1                             0.01859                      0.27402   

           Logistic Regression Precision-Recall  
Accuracy                                0.80733  
Recall                                  0.22733  
Precision                               0.22311  
F1                                      0.22520

* Similar to the training set, the Precision-Recall optimized threshold model has the most balanced performance of the three models.

### Conclusion
- We have built a model that has a ~0.81 recall on the training set which we can use to help predict the response rate of customers.

* The logistic regression models are giving a generalized performance on training and test set.

*  Using the model with default threshold will give a very low recall but high precision score.
* Using the model with 0.13 threshold will give a relatively high recall and higher f1 score, but there will be a loss in accuracy and precision.
* Using the model with 0.24 threshold will give well-balanced scores overall.

* The largest positive correlation exists with those that are self-employed with 0.01963. An increase in this and all other columns with a positive coefficient should lead to an an increase in responses.

* The largest negative coeffient exists for those that are salaried and those with savings accounts with -0.02274 and -0.02233, respectively. All columns with negative coefficients should lead to a decrease in response rate when they increase.

Model building - Decision Tree
- Build the model and comment on the model performance. - Identify the key variables that have a strong relationship with the dependent variable - Comment on model performance



# Decision Tree

In [93]:
# defining a function to compute different metrics to check performance of a classification model built using sklearn
def model_performance_classification_sklearn(model, predictors, target):
    """
    Function to compute different metrics to check classification model performance

    model: classifier
    predictors: independent variables
    target: dependent variable
    """

    # predicting using the independent variables
    pred = model.predict(predictors)

    acc = accuracy_score(target, pred)  # to compute Accuracy
    recall = recall_score(target, pred)  # to compute Recall
    precision = precision_score(target, pred)  # to compute Precision
    f1 = f1_score(target, pred)  # to compute F1-score

    # creating a dataframe of metrics
    df_perf = pd.DataFrame(
        {"Accuracy": acc, "Recall": recall, "Precision": precision, "F1": f1,},
        index=[0],
    )

    return df_perf

In [94]:
def confusion_matrix_sklearn(model, predictors, target):
    """
    To plot the confusion_matrix with percentages

    model: classifier
    predictors: independent variables
    target: dependent variable
    """
    y_pred = model.predict(predictors)
    cm = confusion_matrix(target, y_pred)
    labels = np.asarray(
        [
            ["{0:0.0f}".format(item) + "\n{0:.2%}".format(item / cm.flatten().sum())]
            for item in cm.flatten()
        ]
    ).reshape(2, 2)

    plt.figure(figsize=(6, 4))
    sns.heatmap(cm, annot=labels, fmt="")
    plt.ylabel("True label")
    plt.xlabel("Predicted label")

### Build Decision Tree Model

In [95]:
model = DecisionTreeClassifier(criterion="gini", random_state=1)
model.fit(X_train, y_train)

DecisionTreeClassifier(random_state=1)

#### Checking model performance on training set

In [96]:
confusion_matrix_sklearn(model, X_train, y_train)

In [97]:
decision_tree_perf_train = model_performance_classification_sklearn(
    model, X_train, y_train
)
decision_tree_perf_train

Accuracy  Recall  Precision      F1
0   1.00000 1.00000    1.00000 1.00000

* 0 errors on the training set.
* Perfect performance on training set.

### Visualizing the Decision Tree

In [98]:
feature_names = list(X_train.columns)
print(feature_names)

['BALANCE', 'SCR', 'HOLDING_PERIOD', 'LEN_OF_RLTN_IN_MNTH', 'NO_OF_L_CR_TXNS', 'NO_OF_L_DR_TXNS', 'TOT_NO_OF_L_TXNS', 'NO_OF_BR_CSH_WDL_DR_TXNS', 'NO_OF_ATM_DR_TXNS', 'NO_OF_NET_DR_TXNS', 'NO_OF_MOB_DR_TXNS', 'NO_OF_CHQ_DR_TXNS', 'FLG_HAS_CC', 'AMT_ATM_DR', 'AMT_BR_CSH_WDL_DR', 'AMT_CHQ_DR', 'AMT_NET_DR', 'AMT_MOB_DR', 'AMT_L_DR', 'FLG_HAS_ANY_CHGS', 'AMT_OTH_BK_ATM_USG_CHGS', 'AMT_MIN_BAL_NMC_CHGS', 'NO_OF_IW_CHQ_BNC_TXNS', 'NO_OF_OW_CHQ_BNC_TXNS', 'FLG_HAS_NOMINEE', 'FLG_HAS_OLD_LOAN', 'OCCUPATION_SAL', 'OCCUPATION_SELF-EMP', 'OCCUPATION_SENP', 'GENDER_M', 'GENDER_O', 'ACC_TYPE_SA', 'AGE_BKT_31-35', 'AGE_BKT_36-40', 'AGE_BKT_41-45', 'AGE_BKT_46-50', 'AGE_BKT_<25', 'AGE_BKT_>50']


In [99]:
plt.figure(figsize=(20, 30))
out = tree.plot_tree(
    model,
    feature_names=feature_names,
    filled=True,
    fontsize=9,
    node_ids=False,
    class_names=None,
)
# below code will add arrows to the decision tree split if they are missing
for o in out:
    arrow = o.arrow_patch
    if arrow is not None:
        arrow.set_edgecolor("black")
        arrow.set_linewidth(1)
plt.show()

In [100]:
# Text report showing the rules of a decision tree -

print(tree.export_text(model, feature_names=feature_names, show_weights=True))

|--- NO_OF_L_DR_TXNS <= 5.50
|   |--- FLG_HAS_CC <= 0.50
|   |   |--- NO_OF_L_CR_TXNS <= 16.50
|   |   |   |--- HOLDING_PERIOD <= 12.50
|   |   |   |   |--- NO_OF_ATM_DR_TXNS <= 1.50
|   |   |   |   |   |--- AMT_ATM_DR <= 13300.00
|   |   |   |   |   |   |--- AMT_L_DR <= 1452590.50
|   |   |   |   |   |   |   |--- AMT_L_DR <= 808460.00
|   |   |   |   |   |   |   |   |--- AMT_L_DR <= 804235.00
|   |   |   |   |   |   |   |   |   |--- SCR <= 210.50
|   |   |   |   |   |   |   |   |   |   |--- AMT_BR_CSH_WDL_DR <= 700285.00
|   |   |   |   |   |   |   |   |   |   |   |--- truncated branch of depth 5
|   |   |   |   |   |   |   |   |   |   |--- AMT_BR_CSH_WDL_DR >  700285.00
|   |   |   |   |   |   |   |   |   |   |   |--- truncated branch of depth 2
|   |   |   |   |   |   |   |   |   |--- SCR >  210.50
|   |   |   |   |   |   |   |   |   |   |--- SCR <= 213.50
|   |   |   |   |   |   |   |   |   |   |   |--- truncated branch of depth 2
|   |   |   |   |   |   |   |   |   |   |--- SCR > 

In [101]:
# importance of features in the tree building ( The importance of a feature is computed as the
# (normalized) total reduction of the criterion brought by that feature. It is also known as the Gini importance )

print(
    pd.DataFrame(
        model.feature_importances_, columns=["Imp"], index=X_train.columns
    ).sort_values(by="Imp", ascending=False)
)

                             Imp
SCR                      0.10583
BALANCE                  0.09671
AMT_L_DR                 0.08944
LEN_OF_RLTN_IN_MNTH      0.06515
AMT_ATM_DR               0.06475
AMT_CHQ_DR               0.06297
TOT_NO_OF_L_TXNS         0.05725
NO_OF_L_CR_TXNS          0.05592
AMT_BR_CSH_WDL_DR        0.05336
HOLDING_PERIOD           0.04271
AMT_NET_DR               0.04076
NO_OF_L_DR_TXNS          0.03872
AMT_MOB_DR               0.03028
NO_OF_CHQ_DR_TXNS        0.02072
NO_OF_BR_CSH_WDL_DR_TXNS 0.01810
OCCUPATION_SELF-EMP      0.01666
GENDER_M                 0.01261
FLG_HAS_CC               0.01228
FLG_HAS_OLD_LOAN         0.01161
FLG_HAS_ANY_CHGS         0.00906
NO_OF_ATM_DR_TXNS        0.00901
AGE_BKT_<25              0.00875
AGE_BKT_41-45            0.00869
OCCUPATION_SAL           0.00844
OCCUPATION_SENP          0.00771
NO_OF_NET_DR_TXNS        0.00720
AGE_BKT_46-50            0.00639
ACC_TYPE_SA              0.00634
FLG_HAS_NOMINEE          0.00476
AGE_BKT_>5

In [102]:
importances = model.feature_importances_
indices = np.argsort(importances)

plt.figure(figsize=(8, 8))
plt.title("Feature Importances")
plt.barh(range(len(indices)), importances[indices], color="violet", align="center")
plt.yticks(range(len(indices)), [feature_names[i] for i in indices])
plt.xlabel("Relative Importance")
plt.show()

* Marketing score is the most important feature, followed by the average balance of the customer and the total amount debited.
* This seems to suggest that the marketing score assigned to each customer is quite useful and accurate.
* This also seems to suggest that those with more in their account are more likely to respond.

### Model Improvement

In [103]:
# Choose the type of classifier.
estimator = DecisionTreeClassifier(random_state=1)

# Grid of parameters to choose from
parameters = {
    "max_depth": np.arange(6, 15),
    "min_samples_leaf": [1, 2, 5, 7, 10],
    "max_leaf_nodes": [2, 3, 5, 10],
}

# Type of scoring used to compare parameter combinations
acc_scorer = make_scorer(recall_score)

# Run the grid search
grid_obj = GridSearchCV(estimator, parameters, scoring=acc_scorer, cv=5)
grid_obj = grid_obj.fit(X_train, y_train)

# Set the clf to the best combination of parameters
estimator = grid_obj.best_estimator_

# Fit the best algorithm to the data.
estimator.fit(X_train, y_train)

DecisionTreeClassifier(max_depth=6, max_leaf_nodes=10, random_state=1)

#### Checking performance on training set

In [104]:
confusion_matrix_sklearn(estimator, X_train, y_train)

In [105]:
decision_tree_tune_perf_train = model_performance_classification_sklearn(
    estimator, X_train, y_train
)
decision_tree_tune_perf_train

Accuracy  Recall  Precision      F1
0   0.87514 0.01636    0.87879 0.03212

* We see a very high precision and accuracy in this model, but low recall and f1.

### Visualizing the Decision Tree

In [106]:
plt.figure(figsize=(10, 10))
out = tree.plot_tree(
    estimator,
    feature_names=feature_names,
    filled=True,
    fontsize=9,
    node_ids=False,
    class_names=None,
)
# below code will add arrows to the decision tree split if they are missing
for o in out:
    arrow = o.arrow_patch
    if arrow is not None:
        arrow.set_edgecolor("black")
        arrow.set_linewidth(1)
plt.show()

In [107]:
# Text report showing the rules of a decision tree -

print(tree.export_text(estimator, feature_names=feature_names, show_weights=True))

|--- NO_OF_L_DR_TXNS <= 5.50
|   |--- FLG_HAS_CC <= 0.50
|   |   |--- NO_OF_L_CR_TXNS <= 16.50
|   |   |   |--- weights: [4738.00, 296.00] class: 0
|   |   |--- NO_OF_L_CR_TXNS >  16.50
|   |   |   |--- weights: [626.00, 118.00] class: 0
|   |--- FLG_HAS_CC >  0.50
|   |   |--- weights: [2203.00, 352.00] class: 0
|--- NO_OF_L_DR_TXNS >  5.50
|   |--- OCCUPATION_SELF-EMP <= 0.50
|   |   |--- SCR <= 552.50
|   |   |   |--- weights: [2754.00, 373.00] class: 0
|   |   |--- SCR >  552.50
|   |   |   |--- AMT_ATM_DR <= 7650.00
|   |   |   |   |--- weights: [370.00, 49.00] class: 0
|   |   |   |--- AMT_ATM_DR >  7650.00
|   |   |   |   |--- LEN_OF_RLTN_IN_MNTH <= 182.50
|   |   |   |   |   |--- weights: [609.00, 260.00] class: 0
|   |   |   |   |--- LEN_OF_RLTN_IN_MNTH >  182.50
|   |   |   |   |   |--- weights: [216.00, 30.00] class: 0
|   |--- OCCUPATION_SELF-EMP >  0.50
|   |   |--- BALANCE <= 42128.49
|   |   |   |--- TOT_NO_OF_L_TXNS <= 56.50
|   |   |   |   |--- weights: [129.00, 90.00]

**Observations**
* The tree has simplified significantly from the original.
* Overall model performance has decreased significantly, but the scores are still relatively good and useable.

In [108]:
# importance of features in the tree building ( The importance of a feature is computed as the
# (normalized) total reduction of the criterion brought by that feature. It is also known as the Gini importance )

print(
    pd.DataFrame(
        model.feature_importances_, columns=["Imp"], index=X_train.columns
    ).sort_values(by="Imp", ascending=False)
)

                             Imp
SCR                      0.10583
BALANCE                  0.09671
AMT_L_DR                 0.08944
LEN_OF_RLTN_IN_MNTH      0.06515
AMT_ATM_DR               0.06475
AMT_CHQ_DR               0.06297
TOT_NO_OF_L_TXNS         0.05725
NO_OF_L_CR_TXNS          0.05592
AMT_BR_CSH_WDL_DR        0.05336
HOLDING_PERIOD           0.04271
AMT_NET_DR               0.04076
NO_OF_L_DR_TXNS          0.03872
AMT_MOB_DR               0.03028
NO_OF_CHQ_DR_TXNS        0.02072
NO_OF_BR_CSH_WDL_DR_TXNS 0.01810
OCCUPATION_SELF-EMP      0.01666
GENDER_M                 0.01261
FLG_HAS_CC               0.01228
FLG_HAS_OLD_LOAN         0.01161
FLG_HAS_ANY_CHGS         0.00906
NO_OF_ATM_DR_TXNS        0.00901
AGE_BKT_<25              0.00875
AGE_BKT_41-45            0.00869
OCCUPATION_SAL           0.00844
OCCUPATION_SENP          0.00771
NO_OF_NET_DR_TXNS        0.00720
AGE_BKT_46-50            0.00639
ACC_TYPE_SA              0.00634
FLG_HAS_NOMINEE          0.00476
AGE_BKT_>5

In [109]:
importances = estimator.feature_importances_
indices = np.argsort(importances)

plt.figure(figsize=(8, 8))
plt.title("Feature Importances")
plt.barh(range(len(indices)), importances[indices], color="violet", align="center")
plt.yticks(range(len(indices)), [feature_names[i] for i in indices])
plt.xlabel("Relative Importance")
plt.show()

* Decision tree after pre-pruning has simplified greatly, but the feature importances have changed.
* The most important features are now number of debit transactions and the dummy category of self employed individuals.
* The self-employed feature importance falls in line with some of our earlier observations
* The number of debit transactions seems to correlate logically, those that make debit transactions more often would seem to have higher response rates to this campaign.
* Original decision tree seems to be the best model of the two, overall.

**Observations from decision rules**
* Most columns have little importance to response rate.
* Self-employed individuals, those with higher marketing scores, and those with higher average balances will respond more often.
* Target markets would be self-employed individuals, those with more capital, those with higher marketing scores, and those with many debit transactions.
* Ideal target would be those that satisfy multiple important features.

* **Cost Complexity Pruning**

In [110]:
clf = DecisionTreeClassifier(random_state=1)
path = clf.cost_complexity_pruning_path(X_train, y_train)
ccp_alphas, impurities = path.ccp_alphas, path.impurities

In [111]:
pd.DataFrame(path)

ccp_alphas  impurities
0       0.00000     0.00000
1       0.00004     0.00014
2       0.00005     0.00028
3       0.00005     0.00043
4       0.00005     0.00085
..          ...         ...
477     0.00111     0.21227
478     0.00152     0.21379
479     0.00154     0.21533
480     0.00233     0.21766
481     0.00355     0.22121

[482 rows x 2 columns]

In [112]:
fig, ax = plt.subplots(figsize=(10, 5))
ax.plot(ccp_alphas[:-1], impurities[:-1], marker="o", drawstyle="steps-post")
ax.set_xlabel("effective alpha")
ax.set_ylabel("total impurity of leaves")
ax.set_title("Total Impurity vs effective alpha for training set")
plt.show()

In [113]:
clfs = []
for ccp_alpha in ccp_alphas:
    clf = DecisionTreeClassifier(random_state=1, ccp_alpha=ccp_alpha)
    clf.fit(X_train, y_train)
    clfs.append(clf)
print(
    "Number of nodes in the last tree is: {} with ccp_alpha: {}".format(
        clfs[-1].tree_.node_count, ccp_alphas[-1]
    )
)

Number of nodes in the last tree is: 1 with ccp_alpha: 0.0035450069307925436


In [114]:
clfs = clfs[:-1]
ccp_alphas = ccp_alphas[:-1]

node_counts = [clf.tree_.node_count for clf in clfs]
depth = [clf.tree_.max_depth for clf in clfs]
fig, ax = plt.subplots(2, 1, figsize=(10, 7))
ax[0].plot(ccp_alphas, node_counts, marker="o", drawstyle="steps-post")
ax[0].set_xlabel("alpha")
ax[0].set_ylabel("number of nodes")
ax[0].set_title("Number of nodes vs alpha")
ax[1].plot(ccp_alphas, depth, marker="o", drawstyle="steps-post")
ax[1].set_xlabel("alpha")
ax[1].set_ylabel("depth of tree")
ax[1].set_title("Depth vs alpha")
fig.tight_layout()

### Recall vs alpha for training and testing sets

In [115]:
recall_train = []
for clf in clfs:
    pred_train = clf.predict(X_train)
    values_train = recall_score(y_train, pred_train)
    recall_train.append(values_train)

recall_test = []
for clf in clfs:
    pred_test = clf.predict(X_test)
    values_test = recall_score(y_test, pred_test)
    recall_test.append(values_test)

In [116]:
fig, ax = plt.subplots(figsize=(15, 5))
ax.set_xlabel("alpha")
ax.set_ylabel("Recall")
ax.set_title("Recall vs alpha for training and testing sets")
ax.plot(ccp_alphas, recall_train, marker="o", label="train", drawstyle="steps-post")
ax.plot(ccp_alphas, recall_test, marker="o", label="test", drawstyle="steps-post")
ax.legend()
plt.show()

In [117]:
index_best_model = np.argmax(recall_test)
best_model = clfs[index_best_model]
print(best_model)

DecisionTreeClassifier(random_state=1)


* Original Decision Tree model is determined to be the best of the two based on this return.
* The pruning model should expect to see worse performance when compared to the original.

### Let's check the performance on test set

**Using the decision tree with default parameters**

In [118]:
confusion_matrix_sklearn(model, X_test, y_test)

In [119]:
decision_tree_perf_test = model_performance_classification_sklearn(
    model, X_test, y_test
)
decision_tree_perf_test

Accuracy  Recall  Precision      F1
0   0.93833 0.78078    0.73503 0.75722

**Using the hyperparameter tuned decision tree**

In [120]:
confusion_matrix_sklearn(estimator, X_test, y_test)

In [121]:
decision_tree_tune_perf_test = model_performance_classification_sklearn(
    estimator, X_test, y_test
)
decision_tree_tune_perf_test

Accuracy  Recall  Precision      F1
0   0.87833 0.01353    0.90909 0.02667

### Comparing Decision Tree models

In [122]:
# training performance comparison

models_train_comp_df = pd.concat(
    [decision_tree_perf_train.T, decision_tree_tune_perf_train.T], axis=1,
)
models_train_comp_df.columns = ["Decision Tree sklearn", "Decision Tree (Pre-Pruning)"]
print("Training performance comparison:")
models_train_comp_df

Training performance comparison:


Decision Tree sklearn  Decision Tree (Pre-Pruning)
Accuracy                 1.00000                      0.87514
Recall                   1.00000                      0.01636
Precision                1.00000                      0.87879
F1                       1.00000                      0.03212

In [123]:
# testing performance comparison

models_test_comp_df = pd.concat(
    [decision_tree_perf_test.T, decision_tree_tune_perf_test.T], axis=1,
)
models_test_comp_df.columns = ["Decision Tree sklearn", "Decision Tree (Pre-Pruning)"]
print("Test set performance comparison:")
models_test_comp_df

Test set performance comparison:


Decision Tree sklearn  Decision Tree (Pre-Pruning)
Accuracy                 0.93833                      0.87833
Recall                   0.78078                      0.01353
Precision                0.73503                      0.90909
F1                       0.75722                      0.02667

### Conclusion
* The pre-pruning model has worse performance on the training set in every metric.
* The pre-pruning model has higher precision, but lower in all other metrics, on the test set.

# Insights and Recommendations

* We have been able to build multiple predictive models of response rates with good performance. We have been able to compare them and have identified the pros and cons of each of the competitive models for the business to use.

* Across the models we have identified that the most important features are the marketing scores, balances, number of debit transaction, and self-employed customers.

* The marketing scores have been shown to be relatively accurate in predicting response, I would advise the business to focus on this scoring and seek to improve its accuracy for further modeling.

* Customers with higher capital and those that make more debit transactions are more likely to respond, and should therefore be targeted in marketing and sales.

* Customers that have a credit card are also shown to have a higher response rate.